### DISCUSSION TITLE
rfsrc and HPC - LSF exit code 139.

### DISCUSSION THREAD

**Discussion by cecelg-usc** [kogalur/randomForestSRC#357] [Q&A]
Hi All,

I am using rfsrc on a large, multi-year geoscience dataset (15GB) to understand driver and response relationships (primarily using VIMP lists and partial dependency plots). I have successfully run rfsrc on smaller versions of my dataset locally and on an **HPC (high memory non/parallel) cluster** though my university. However, when I try to use my full dataset, I get an error from my **LSF job**, **exited with exit code 139.**. This seems to be a segmentation issue? Or a memory issue on the HPC side. However, after speaking the HPC system admin, they think that given my memory allocation, the exit code most likely is not reflecting a memory issue on their end. 

**So my question is has anyone had this issue before?** And what could potentially cause it? Am I using the correct HPC queue? Is the bigmem queue ideal if I have used dectect.cores and rf.cores with rfsrc? Or is the multiple core use causing segmentation? And finally, should I be using the parallel queue instead? See code and BSUBs info below.

Any and all help appreciated. And apologies if this is a naïve question or posted incorrectly.

Ceci

Additional set up info:
lsf job - 1 cpu, 200 Gb of memory allocated 
using R module with R /4.1.0

https://github.dev/cecelg-um/rfsrc-TSS/blob/13a4c9d54d84187bd1023aa9af0be6ae3634691e/rfsrc_example_2003_2020.R#L1-L77

**Comment by kogalur** [comment]
I think you have to convince us with a simple example, that you have everything on the Pegasus side configured correctly and are actually entering and exiting our R-package without error with that example.  The last time we did anything on Pegasus was 2015, so it's been a while, and I recall we had a challenging time getting resources allocated correctly.  You have to be super careful when using code that uses OpenMP.   Here are some questions I have that may or may not be relevant:

1. How many cores are you accessing? How are you configuring this on Pegasus?  What does your LSF batch script look like? Why does it seem like you are only accessing one CPU? 
2. If you are convinced that things are working on smaller data sets, perhaps you can subset your data,  reduce ntree, and turn off vimp.

This might be a bit basic, but here is a test script that we used on Pegasus back in the day:

This was our test shell script hello.openmp.sh:

#!/bin/sh
export OMP_NUM_THREADS=24
srun R CMD BATCH --no-save ./hello.openmp.R

This was our R script hello.openmp.R:

dyn.load("hello.openmp.so")
.C("main", 1, "test")

This was our C-code hello.openmp.c:

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

int main (int argc, char *argv[]) {
  int nthreads, threadID;

  // Fork a team of OpenMP threads giving them their own copies of variables.
  #pragma omp parallel private(nthreads, threadID)
  {
    // Get the thread number.
    threadID = omp_get_thread_num();
    nthreads = omp_get_num_threads();
    printf("OpenMP thread = %d of %d reporting for duty.\n",
           threadID,
           nthreads);
    
  }

  printf("\n");
  return 0;
}

To compile the C-code to execute from an R environment, you'll have something like:

gcc-12 -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG   -I/usr/local/include -I/usr/local/include/freetype2 -I/opt/X11/include  -fopenmp -fPIC  -Wall -mtune=core2 -g -O2  -c hello.openmp.c -o hello.openmp.o

gcc-12 -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/Library/Frameworks/R.framework/Resources/lib -L/usr/local/Cellar/gcc/12.2.0/lib/gcc/12/  -o hello.openmp.so hello.openmp.o -fopenmp -F/Library/Frameworks/R.framework/.. -framework R -Wl,-framework -Wl,CoreFoundation

Your paths on Pegasus will probably be different, but you can locally compile RF-SRC and see what the actual paths will be by examining the trace that R CMD BUILD produces from our source package.

So now, the command

./hello.openmp.sh

should execute all of the above and give something like:

OpenMP thread = 1 of 24 reporting for duty.
OpenMP thread = 6 of 24 reporting for duty.
OpenMP thread = 10 of 24 reporting for duty.
...
[[1]]
[1] 1

[[2]]
[1] "test"

The above may or may not be useful to you.  As an alternative, while I understand you want to use the institutional resources available to you, we have found that accessing cloud services like Google, or Amazon, is a lot easier.  You set up your RAM and CPU requirements.  You virtualize the machine. You install RF-SRC, and it will run out of the box.

**Comment by kogalur** [comment]
As a follow up, I see that you say you have subsetted your data and run your example successfully.  In big-data scenarios, it's crucial to know how many threads are spawned, whether you have missing data (we recommend you pre-impute if you do), and to turn importance OFF during the grow cycle.  It's much better to get vimp values in a predict call, AFTER the model has been grown.

**Comment by cecelg-usc** [comment]
@kogalur - thanks for the swift response. As you suggest, I think it has something to do with my Pegasus configuration. However, the error occurred when I tried rfsrc using Rsrcipt myfile.R in a serial job on the "bigmem" queue. As you note this is 1 cpu with 200 Gb of memory allocated. My hope was to get this set up working before trying to make my job parallel. So I simply took the OpenMP: shared memory parallel processing workflow (https://www.randomforestsrc.org/articles/hybrid.html) from local script (loaded the parallel package and used rf.cores = x) then used it on a batch job on Pegasus. This is where I got the exit code 139 but it was unexpected since I did not explicitly do any parallel processing. Does this happen regardless of me using the parallel queue or not?. I believe this answers (1). (2) is a what I kind of have defaulted to, running rfsrc on temporal data subsets by year (2003-2009, then 2009-2020).  However, I rather not do this, but have it on the full extent at once since the VIMPs slightly change over the decades. 

My LSF batch script looks like this: 

#!/bin/bash
#BSUB -J bigmem_200Gb_1CPU_2003_2020
#BSUB -P tss_pegasus
#BSUB -o %J.out
#BSUB -e %J.err
#BSUB -W 10:00
#BSUB -q bigmem
#BSUB -n 1
#BSUB -R "rusage[mem=200000]"
#BSUB -B
#BSUB -N
#BSUB -u cxl1148@miami.edu
#
Rscript rfsrc_TSS_2003_2020.R

**Comment by cecelg-usc** [comment]
I also want to say thanks for the test shell script. I will attempt to give it a go, however, I am new to most of this. So I wanted to make sure I understand what the test script does first. You have a shell script that calls a R script which calls a C script? Then you separately have to compile the c script to be readable in R? I have never done the latter or used gcc knowingly so I will try and figure this out and report back.

I am unsure if they had this before, but Pegasus now has gcc/8.3.0 compiler modules with openmpi/4.0.5 and R/4.1.0 included. Do you think this provides a work around for the replica and harness scripts that you use here: https://www.randomforestsrc.org/articles/hybrid.html. Would these harness scripts still be needed for parallel processing if the openmpi module is loaded? And lastly is any of this necessary when simply using an LSF serial job? Should segmentation even occur if we are not doing things in parallel?

**Comment by kogalur** [comment]
I think you need to take a step back and find out what resources you are allocating and actually accessing on that system when you submit a batch job. A good approach is to develop and run the test scripts on your local machine, convince yourself that they are working, and then wrap them in a Pegasus shell script for submission to the supercomputer.

You are correct with the workflow:

bash --> R --> C

Your first bash script could simply call the R script on your local machine or your home environment on Pegasus. Don't even bother with the #BSUB and submission process. I just ran it on my desktop here as an example.  You will need to adjust your path information for the GCC compiler.

You can confirm this using

> which gcc

You can also just try to install RF-SRC in your home environment using

> R CMD INSTALL randomForestSRC

using our source code.  You will be able to see all the compiler paths in the trace output of this command.  

Also note that we use OpenMP, not OpenMPI.  But this shouldn't matter to you, as we don't spawn any MPI processes while OpenMP threads are active. That note you mentioned above was a simple example demonstrating that it is possible given very specific architectures to do that.  As it stands, our package is not designed to spread a forest over MPI processes, each of which contain an OpenMP process (as shown in the diagram on that page), and come back with coherent results.

If n = 5.3M, and p = 24, I'm really surprised that you would need 200GB of RAM to build the model.  This assumes VIMP is off. Maybe you'd have to turn some other options off as well, though that's not clear at this point. That array might cost you about 1GB on the C-side, that's it. But the total cost will be more, because we have to build and store the model, of course.

You really want to make sure you don't have missing data. Make sure na.impute is turned off.  Imputing on the fly with big data is a bad idea.

**Comment by cecelg-usc** [comment]
Here are my .sh, .R, .c, .o, and .so files: https://github.com/cecelg-um/rfsrc-TSS/commit/b05dce969ae2d318406475aad3d05dd7739da57f

my pegasus path is as follows and confirmed it was the same used during the randomForestSRC install: /share/builds/spack-compilers/opt/spack/linux-centos7-sandybridge/gcc-8.3.1/gcc-8.3.0-mnwujmnexs6zvenuhlu5m4io2om4sg55/bin/gcc

hence my compile commands were as follows: 
/share/builds/spack-compilers/opt/spack/linux-centos7-sandybridge/gcc-8.3.1/gcc-8.3.0-mnwujmnexs6zvenuhlu5m4io2om4sg55/bin/gcc -I"/share/builds/spack/opt/spack/linux-centos7-sandybridge/gcc-8.3.0/r-4.1.0-wsuhybx36dpwkrryfjhg3kph2ytyplj7/rlib/R/include" -DNDEBUG   -I/usr/local/include  -fopenmp  -fpic  -g -O2  -c hello.openmp.c -o hello.openmp.o

/share/builds/spack-compilers/opt/spack/linux-centos7-sandybridge/gcc-8.3.1/gcc-8.3.0-mnwujmnexs6zvenuhlu5m4io2om4sg55/bin/gcc -shared -L/share/builds/spack/opt/spack/linux-centos7-sandybridge/gcc-8.3.0/r-4.1.0-wsuhybx36dpwkrryfjhg3kph2ytyplj7/rlib/R/lib -L/share/builds/spack/opt/spack/linux-centos7-sandybridge/gcc-8.3.0/r-4.1.0-wsuhybx36dpwkrryfjhg3kph2ytyplj7/rlib/R/lib -Wl,-rpath,/share/builds/spack/opt/spack/linux-centos7-sandybridge/gcc-8.3.0/r-4.1.0-wsuhybx36dpwkrryfjhg3kph2ytyplj7/rlib/R/lib -o hello.openmp.so hello.openmp.o -fopenmp -L/share/builds/spack/opt/spack/linux-centos7-sandybridge/gcc-8.3.0/r-4.1.0-wsuhybx36dpwkrryfjhg3kph2ytyplj7/rlib/R/lib -lR

no errors here. however.... when I try and execute the shell I get the following....

(base) [cxl1148@login4 tss_pegasus]$ ./hello.openmp.sh
./hello.openmp.sh: line 3: srun: command not found

not sure what I did wrong here but I am at a bit of an impasse. note I have not been able to install Rmpi yet... could this be the issue?

**Comment by cecelg-usc** [comment]
as far as missing data is concerned, I use na.omit on my dataframe beforehand which drop n to 5.3 M before training. Before that is roughly 100x larger. Also, I think I have VIMP on when training (https://github.com/cecelg-um/rfsrc-TSS/blob/main/rfsrc_example_2003_2020.R)? Ln 51 below:

 rf_overall[[j]]=rfsrc(as.formula(paste0(outcomes[j],'~.')),data=train[,-match(c(removevar, outcomes[-j]), names(train))],importance = T)

In terms of RAM use, when I use the graphic interface of R on pegasus, I see my RAM usage grow from 15Gb initially (before using na.omit and freeing up extra space), to well into the 100s of Gb. Using the LSF and allocating 200 Gb of mem, I get only 3.5 Gb of unused memory (delta memory) when the exit code terminates my job.

**Comment by ishwaran** [comment]
You need to run a lower memory profile forest as I had suggested earlier.  Try the following:

`rf_overall[[j]]=rfsrc(as.formula(paste0(outcomes[j],'~.')),data=train[,-match(c(removevar, outcomes[-j]), names(train))],ntree=50,sampsize=50000,nodesize=500,perf.type="none")`

**Reply by cecelg-usc** [reply]
@ishwaran - Thanks again for the advice. I am running this now on pegasus, but in serial.... so it takes a moment.... Will report back as soon as I can!

I also noticed you dropped `importance = T` in my code. @kogalur already eluded to calculating the VIMP later. Its seems that not doing so during forest growth caused my script to crash later on since I don't get and R2 or OOB information in my model off the bat.

I will just see if this lower memory profile does not create the segmentation error on the full extent, but I wanted to see the relative model performances first on a smaller extent. As of now I cannot check.

**Reply by cecelg-usc** [reply]
@ishwaran  - model ran very quickly, thank you for that. However, I wanted to see how it performed against the higher memory profile forest that rfsrc defaulted to originally. However, as I mention, I am having trouble finding the R2  and OOB error. I was able to call vimp after the training, but cannot seem to find the $err.rate on the new rf_overall variable now.

**Comment by kogalur** [comment]
Just get rid of the "srun" in the line. I inadvertently left that in. You don't need it if you are running it locally.  Remember, the R-code in that shell script can be executed directly from an R command prompt in an R-environment shell.  We just wrap it in a batch command for convenience.  If you are running it on a system like Pegasus that only accepts batch commands, you have will likely need to prepend it with the relevant batch submission protocols.

Stay away from anything MPI for now.  We don't need it.

**Comment by cecelg-usc** [comment]
@kogalur got this to work. I noticed before that for some reason I don't have success R CMD BATCH but rather Rscript in my shell commands and lsf jobs. So I had to alter the hello.openmp.sh to the following:

#!/bin/sh
export OMP_NUM_THREADS=24
Rscript --no-save ./hello.openmp.R

also had permissions issues so had to execute the shell using:

sh hello.openmp.sh

rather than ./hello.openmp.sh

Finally got the following:

OpenMP thread = 2 of 24 reporting for duty.
OpenMP thread = 8 of 24 reporting for duty.
OpenMP thread = 11 of 24 reporting for duty.
OpenMP thread = 15 of 24 reporting for duty.
OpenMP thread = 3 of 24 reporting for duty.
OpenMP thread = 21 of 24 reporting for duty.
OpenMP thread = 19 of 24 reporting for duty.
OpenMP thread = 0 of 24 reporting for duty.
OpenMP thread = 1 of 24 reporting for duty.
OpenMP thread = 20 of 24 reporting for duty.
OpenMP thread = 5 of 24 reporting for duty.
OpenMP thread = 22 of 24 reporting for duty.
OpenMP thread = 23 of 24 reporting for duty.
OpenMP thread = 6 of 24 reporting for duty.
OpenMP thread = 7 of 24 reporting for duty.
OpenMP thread = 9 of 24 reporting for duty.
OpenMP thread = 10 of 24 reporting for duty.
OpenMP thread = 4 of 24 reporting for duty.
OpenMP thread = 12 of 24 reporting for duty.
OpenMP thread = 18 of 24 reporting for duty.
OpenMP thread = 17 of 24 reporting for duty.
OpenMP thread = 16 of 24 reporting for duty.
OpenMP thread = 14 of 24 reporting for duty.
OpenMP thread = 13 of 24 reporting for duty.

[[1]]
[1] 1

[[2]]
[1] "test"

Seems to be working!

So does this mean that openmp is communicating with my R script? If so, would it be safe to assume the same if I set up my script on a parallel queue? Are there any additional steps, like compiling some C script, that I need to do before submitting some job? And lastly, do you think I should regardless set up a more low mem profile as @ishwaran suggests to avoid such segmentation errors that were at the heart of this post?

**Reply by kogalur** [reply]
Yes, you are accessing 24 cores using OpenMP. If you submit your primary job the same way you executed the above script, you should access the same number of cores.  If you ran this locally on your node, and not as a batch job on the parallel queue, you will need to make a batched parallel queue version of your script.  This would involve adding those #BSUB commands in your queue.  I don't think you'll need to compile RF-SRC. It depends on whether the queued version uses a shared R-library and extensions (CRAN packages) or the library on your local node.  I'm not sure how your system is set up.  Then run RF-SRC using a low memory profile.  You'll be able to tell if you are running in parallel by putting adding a proc.time() at the end of your R-script, and seeing dramatic differences between serial and parallel execution.
